misc/pytorch_toolkit/breast_ultrasound_simulation/src/solve.py

```diff
@@ -0,0 +1,358 @@
+                        self.writer.add_scalar('Loss/Generator',
+                                               self.generator_adversarial_loss.item(),
+                                               i_batch + len(self.train_data) * epoch)
+                        self.writer.add_scalar('Loss/Reconstruction',
+                                               self.reconstruction_loss.item(),
+                                               i_batch + len(self.train_data) * epoch)
+                        self.writer.add_scalar('LR', get_lr(
+                            self.g_opt), i_batch + len(self.train_data) * epoch)
+
+                    if(i_batch == 0 and epoch % self.args.vis_step == 0):
+                        self.writer.add_image('Fake', torchvision.utils.make_grid(self.fake[:16].detach(
+                        ).cpu(), scale_each=True), i_batch + len(self.train_data) * epoch)
+                        self.writer.add_image('Real', torchvision.utils.make_grid(self.real[:16].detach(
+                        ).cpu(), scale_each=True), i_batch + len(self.train_data) * epoch)
+                        self.writer.add_image('stage0', torchvision.utils.make_grid(
+                            X[:16].detach().cpu(), scale_each=True), i_batch + len(self.train_data) * epoch)
+                    self.writer.flush()
+
+            state_dict = {
+                "generator1_weights": self.gen.state_dict(),
+                "discriminator1_weights": self.dis.state_dict(),
+                "epoch": epoch,
+                "generator1_optimizer": self.g_opt.state_dict(),
+                "discriminator1_optimizer": self.d_opt.state_dict()}
+            if(epoch % 2 == 0):
+                torch.save(state_dict, os.path.join("../checkpoints/",
+                           args.name + "/" + str(epoch % 20) + ".pt"))
+            torch.save(
+                state_dict,
+                os.path.join(
+                    "../checkpoints/",
+                    args.name +
+                    "/" +
+                    "latest" +
+                    ".pt"))
+
+    def test(self):
+        self.gen.eval()
+        print("Starting Testing")
+        avg_time = 0
+        for i_batch, sample in enumerate(self.test_data):
+            X, _, n = sample
+            X = X.to(device)
+
+            # generating fake images
+            with torch.no_grad():
+                start_time = time.time()
+                self.fake = self.gen(X)
+                total_time = time.time() - start_time
+            avg_time += total_time
+
+            for i in range(X.shape[0]):
+                torchvision.utils.save_image(
+                    self.fake[i].detach().cpu(), os.path.join(
+                        self.output_dir, n[i]), normalize=True, scale_each=True)
+
+            print('Step:{}'.format(i_batch))
+
+
+class solver_inter2d(solver):
+    def __init__(self, args, train_data=None, test_data=None, restore=1):
```

Comment:
You inherit `solver` class but do not call its `__init__` method:
`super().__init__(args, train_data=None, test_data=None, restore=1)`
Adding this call would allow not to duplicate common code. 
You will need to extract common part of all successor classes (`solver_inter2d` and `solver_inter3d`) into `solver`'s `__init__` method and add to the successors' `__init__`'s only differing functionality. This would make code more readable, compact and easier to maintain. [Learn more](https://docs.python.org/3/library/functions.html#super) about python's super.